Add wavefront obj and x3d export

tu500 opened the pull request:
TODO:
* [ ] Code review
* [ ] Add a few test cases
* [ ] This PR exhibits the same export issues as AMF and OFF. If merged, we should create new issues tracking this for OBJ and X3D as well, referencing the two mentioned issues: #909 and #910

kintel commented:
@donbright Didn't you explore OBJ export/import a while back? Could you take a quick look at how this compares to your explorations?

kintel commented on Info.plist:
```diff
@@ -93,6 +93,26 @@
+			<key>CFBundleTypeName</key>
+			<string>X3D file</string>
+			<key>CFBundleTypeRole</key>
+			<string>Viewer</string>
```
We cannot import X3D files, so this entire block should be removed. Same for OBJ.

kintel commented on src/MainWindow.ui:
```diff
@@ -1288,6 +1290,24 @@
+  <action name="fileActionExportX3D">
+   <property name="icon">
+    <iconset resource="../openscad.qrc">
+     <normaloff>:/images/X3D.png</normaloff>:/images/X3D.png</iconset>
```
This image resource is missing. Same for `OBJ.png`

tu500 replied:
Do you have a way to generate new icons, a script? Where do the existing come from?

kintel replied:
They are hand-drawn. There are some SVGs buried in a PR somewhere (#1214). I don't think icons are necessary unless these icons make it into a toolbar.

kintel commented on src/export_indexed_triangles.cc:
```diff
@@ -0,0 +1,128 @@
+/*
+ *	OpenSCAD (www.openscad.org)
+ *	Copyright (C) 2009-2011 Clifford Wolf <clifford@clifford.at> and
+ *													Marius Kintel <marius@kintel.net>
```
Indentation glitch

kintel commented on src/export_indexed_triangles.cc:
```diff
@@ -0,0 +1,128 @@
+	std::vector<export_indexed_triangle> tris;
+
+	for(const auto &p : triangulated.polygons) {
+		assert(p.size() == 3); // STL only allows triangles
```
Comment is misleading

kintel commented on src/export_indexed_triangles.h:
```diff
@@ -0,0 +1,35 @@
+#include "polyset.h"
+
+
+struct export_vec3 {
```
Why not use `Eigen::Vector3d` as we do other places, see `linalg.h`

kintel commented on src/export_indexed_triangles.h:
```diff
@@ -0,0 +1,35 @@
+	}
+};
+
+struct export_indexed_triangle {
```
We could probably use `IndexedTriangle` here. See `GeometryUtils.h`

kintel commented on src/export_indexed_triangles.cc:
```diff
@@ -0,0 +1,128 @@
+#include "cgalutils.h"
+
+
+static size_t insert_vector(std::vector<export_vec3> &vecs, const export_vec3 &vec)
```
This is not very performant. A better idea could be to use `Reindexer` for this; see 'Reindexer.h'.
`Reindexer` is used multiple other places in the codebase.

kintel commented on src/export_obj.cc:
```diff
@@ -0,0 +1,123 @@
+		auto r = export_indexed_triangles(*ps);
+		append_obj(r.first, r.second, output);
+	}
+	// else if (const Polygon2d *poly = dynamic_cast<const Polygon2d *>(geom.get())) {
```
Any reasons why this was commented out? It looks resonable.

kintel replied:
I think this should simply be re-enabled then.

tu500 commented on src/export_obj.cc:
```diff
@@ -0,0 +1,123 @@
+		auto r = export_indexed_triangles(*ps);
+		append_obj(r.first, r.second, output);
+	}
+	// else if (const Polygon2d *poly = dynamic_cast<const Polygon2d *>(geom.get())) {
```
Yes it is, I just got confused about the capabilities of obj files, sorry.

tu500 commented:
Considering test cases, I don't know enough about the code base to generate good test cases and I didn't find any tests for the other exports. If output is deterministic one could do something like

```sh
echo 'cube(10);' > test.scad
openscad -o test.x3d test.scad
diff test.x3d expected.x3d
```

but that doesn't seem right.

Regarding the export issues you mentioned: I do have an overload for PolySets. Do you mean something else by that?

kintel commented:
Tests: It's tricky to test without implementing the corresponding import functionality. I think our current tests rely on bidirectional export/import. A naive check assuming deterministic test output for a very simple testfile is better than no test at all. Note: All our tests are black-box regression tests, so they are pretty easy to implement. See here: https://github.com/openscad/openscad/blob/master/doc/testing.txt

kintel commented:
Export issues: The challenge is that these exports, as far as I can tell, convert directly from a Nef_3 polyhedron into a CGAL Polyhedron, then into the correct file format. This conversion fails in some corner cases. If you look at the e.g. STL export vs. AMF export, you can see that the code path is quite different. The mentioned issues are tracking this discrepancy. I'm not expecting you to fix that, but we should add issues so we know to deal with it later.

..of course, if you decide to fix it that would be awesome :)

tu500 commented:
Ok, this might be getting off topic. As far as I understand it the STL code converts Nef polyhedrons to PolySets and exports those. There is also a code path which converts Nef polyhedrons to CGAL polyhedrons and exports those, but this one is unused. The AMF code converts PolySets to Nef polyhedrons, those to CGAL polyhedrons and then exports them.

The point I wanted to make was that I copied the STL workflow.

tu500 commented:
Also, thanks for the review and especially for explicitly pointing to the right headers. I redid the code and addressed your comments. I'll look into test code and icons.

MichaelAtOz commented:
@tu500 What did you create the icons with?

tu500 commented:
@MichaelAtOz As @kintel mentioned there is a PR with SVG icons #1214.

tu500 commented:
I should mention I added these icons, together with a conversion script, into my svg-icons branch. @kintel If the icons are ok, maybe you want to merge those into the PR, for reference.

Also, the new icons look minimally different than the old ones. If this is a problem I would suggest recreating all icons from the svg version.

kintel commented on src/export_x3d.cc:
```diff
@@ -0,0 +1,150 @@
+		auto r = export_indexed_triangles(*ps);
+		append_x3d(r.first, r.second, output);
+	}
+	// else if (const Polygon2d *poly = dynamic_cast<const Polygon2d *>(geom.get())) {
```
This check should be re-enabled.

kintel commented:
@donbright You worked on smth. similar a while back. I feels like this could be merged early and deal with import as a separate task. Opinions?

kintel commented:
@tu500 ..and there's a merge conflict with `openscad.cc`..

jgwinner commented:
I found a need for OBJ or GLTF export, and found this open ticket ...

Looks like it was a conflicted merge? What's the  status?

I was trying to build a parametric object that could be exported to GLTF or OBJ, for inclusion in a Mixed Reality Extension (Node.JS). 

        == John ==

t-paul commented:
I guess the current status would be abandoned, but it might not be too complicated to revive it. As listed in the task list, some test cases would be needed too (basically integration with the existing test framework).

kintel commented:
Closing as stalled. OBJ import/export has since been merged
